### PR TITLE
Makes grammar more good

### PR DESCRIPTION
**Pull request by Tacolizard**
:cl: Tacolizard
add: When saying question, the game will attempt to detect that it is a question and automatically append a question mark
/:cl:

### CODE REVIEW

**Comment by ShadowDeath6** [comment]
does this only detect if the question word is at the front of the sentence
if not, "I wonder what's going on this shift" and other stuff will be turned into questions incorrectly

**Comment by Tacolizard** [comment]
@ShadowDeath6 don't worry I'm not a stupidhead

it's only detects the first word

**Comment by TechnoAlchemisto** [comment]
When the ERT gets here, we'll all be okay!

File: code/modules/mob/living/say.dm [hunk 1]
```diff
@@ -37,36 +37,36 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	//kinda localization -- rastaf0
 	//same keys as above, but on russian keyboard layout. This file uses cp1251 as encoding.
 	// Location
-	"�" = "right hand",
-	"�" = "left hand",
-	"�" = "intercom",
+	"ê" = "right hand",
+	"ä" = "left hand",
```
Comment (Iamgoofball) [inline code/modules/mob/living/say.dm]: :-1: you broke the encoding

Comment (Tacolizard) [inline code/modules/mob/living/say.dm]: I'm pretty sure I didn't.

I broke the encoding and then fixed it and the diff on github is just flipping out.

**Comment by Tacolizard** [comment]
@DoctoreMad 

why downvote pls

**Comment by outbools** [comment]
@Tacolizard he's shitposter ignore him

**Comment by BeeSting12** [comment]
not a fan of this, players should have to insert question marks themselves and there's situations where you can start a sentence with those words and it not be a question

**Comment by Tacolizard** [comment]
@BeeSting12 what situation is that? I honestly can't think of any

The game already autocapitalizes as well

**Comment by Armhulen** [comment]
this ruins my 'what' response :-1:

**Comment by RemieRichards** [comment]
I don't like this, purely because the implementation is bad.

**Comment by Dorsisdwarf** [comment]
When shouldn't be included since it can, as stated, be a non-question  (not that I like this pr anyway)

**Comment by Tacolizard** [comment]
@RemieRichards I'm going to close this PR, but just for the future can you help me understand the problem with the implementation?

**Comment by RemieRichards** [comment]
Certainly.
For starters, you're parsing mixed case text, but only providing lowercase matching words (you may have been confused by the line above your autoquestion call that capitalises things, but what if the user ALREADY capitalised their text?)

Next is how you handle apostrophes, simply matching against 'what' and 'what's' isn't prefect (for example, what if someone wrote 'whats'? SS13 grammar is already bad enough, so you're missing a whole class of words (whats, whos, wheres, hows (this one isn't even a typo, it's a word!), whys)) To match all 3 versions of each word (word, words, word's) it would be best to use a regex, like: "who'?s?", this will match: who, whos, who's AND an extra, potentially unwanted (But fixable with a different regex) who' (I would leave this one in personally, since people typo all the same)

Thirdly, you use a ```letters``` list, created from ```splittext(the_message, "")```, this doesn't actually work and myself and mso have complained to lummox, we did come up with an alternative though that does work, ```string2charlist(your_text)```

Fourthly, you don't add a ? if there's already one, or a fullstop, except you don't do the same for !, meaning you always get an interobang (?!) even if someone just wanted to write "WHAT!!!" they'd be forced to say "WHAT?!!!"

Fifthly, you loop over the list, comparing the first item in words with the question words, if it matches, you prepare a "?" to add, but then you keep looping over the rest of the questionwords unnecessarily (if the word is 'who' it can't also be 'what'!)

Sixthly, you use ```jointext()``` to add your question mark onto the sentence, which is unnecessary and won't provide any benefit, simpler to just ```new_msg += "?"``` at the end (I would make the ```QM``` var a bool, it's weird to store a question mark in a variable like that)

I think that about covers it?

**Comment by meeeeme** [comment]
You forgot whom'st'd've'ed

**Comment by DoctoreMad** [comment]
@Tacolizard Because its not really necessary. If people can't write well then they should get their head out of their asses instead of relying on a coder to do it.

**Comment by DoctoreMad** [comment]
And besides all those "grammar changes" like replacing "him" with "her" on /me messages is just clunky bloat, who started it?

**Comment by nicbn** [comment]
Try using the following regex instead:

`^((what|when|why|where|how|who|which)(?!( a | an ))($| the | +(is|are|do|does|did) |'(s|re) )|(did|do|does|are|is|should|could|can|was|were)( not|n't)? +(i|you|they|him|her|he|she|we|us|(every|no|any)(body|one)) |(did|does|are|is|should|could|can|was|were)( not|n't)? +(it|this|them|these|those|the) )`

Use http://regexr.com/ to test it.
Also use the flag i, and the flags m and g for testing.

**Comment by Tacolizard** [comment]
@RemieRichards thank you for that. It really does help. I will try to keep some of that in mind in the future.

**Pull request closed without merging**